static/js/services/turndata.js

```diff
@@ -0,0 +1,177 @@
+							urls = turn.serverMap[turn.preferred[i]].urns;
+							break;
+						}
+					}
+				}
+				if (!urls && turn.first) {
+					urls = turn.first.urns;
+				}
+				turnData.urls = urls;
+			} else if (turn && turn.urls) {
+				// Simple case, single region.
+				turnData.urls = turn.urls
+			} else {
+				// Unknown data.
+				turnData.urls = [];
+			}
+			console.log("TURN servers selected: ", turnData, turn.preferred);
+			service.e.triggerHandler("apply", [turnData]);
+
+			return turnData;
+		};
+
+		service.refresh = function(withGeo) {
+			$timeout.cancel(geoRefresh);
+
+			var turn = service.data;
+			service.e.triggerHandler("refresh", [turn]);
+			if (turn.selected === "auto" && turn.geo_uri) {
+				if (geoPreferred !== null) {
+					// Use existing data.
+					turn.preferred = geoPreferred;
+
+				} else {
+					if (!withGeo) {
+						// Avoid triggering spurious GEO request for fast updates.
+						geoRefresh = $timeout(function() {
+							service.refresh(true);
+						}, 1000);
+						return;
+					}
+
+					// Run Geo request.
+					var nonce = randomGen.random({hex: true});
+					$http({
+						method: "POST",
+						url: turn.geo_uri,
+						headers: {"Content-Type": "application/x-www-form-urlencoded"},
+						data: "nonce="+encodeURIComponent(nonce)+"&username="+encodeURIComponent(turn.username)+"&password="+encodeURIComponent(turn.password),
+						timeout: 30000
+					}).then(function(response) {
+						// success
+						if (turn !== service.data) {
+							// No longer our data.
+							return;
+						}
+						if (response.status === 200) {
+							var data = response.data;
+							if (data.success && data.nonce === nonce) {
+								geoPreferred = turn.preferred = data.geo.prefer;
+								console.log("TURN GEO auto selected: ", turn.preferred);
+							}
+						}
+						service.apply();
+					}, function(response) {
+						// failed
+						if (turn !== service.data) {
+							// No longer our data.
+							return;
+						}
+						console.warn("TURN GEO failed:", response.statis, response);
+						service.apply();
+						$timeout.cancel(ttlTimeout);
+						ttlTimeout = $timeout(function() {
+							// Fast retry.
+							console.warn("TURN GEO failed - refreshing early.")
+							api.sendSelf();
+						}, 45000)
+					})
+				}
+			} else {
+				// Set directly.
+				turn.preferred = [];
+				if (turn.selected) {
+					turn.preferred.push(turn.selected);
+				}
+			}
+			service.apply();
+		};
+
+		service.update = function(turn) {
+			$timeout.cancel(ttlTimeout);
+			if (service.data && service.data.preferred) {
+				// Keep preferred list if there is one.
+				turn.preferred = service.data.preferred;
+			}
+			service.data = turn;
+			service.refresh()
+
+			// Support to refresh TURN data when ttl was reached.
+			if (turn.ttl) {
+				ttlTimeout = $timeout(function() {
+					console.log("TURN TTL reached - sending refresh request.");
+					api.sendSelf();
+				}, turn.ttl / 100 * 90 * 1000);
```

Comment:
Please use a constant instead of the "magic" number `90`. This is the percentage of the TTL when it gets refreshed, right?